[DRAFT] docs: Add Tutorial to protect MCP server with Fusionauth (#4115)
FusionAuth/fusionauth-site: astro/src/content/docs/extend/examples/protecting-mcp-servers.mdx

import InlineUIElement from 'src/components/InlineUIElement.astro';
import {RemoteCode} from '@fusionauth/astro-components';

Building a production Model Context Protocol (MCP) server requires you to consider not just the AI-powered tools and services you expose, but also how to control and track which clients can access them. With FusionAuth, you can add OAuth protection to your MCP server, letting you authenticate clients and enforce per-tool scopes, while maintaining full control with a self-hosted Identity Provider.

Starting with a simple, unprotected MCP server and adding OAuth protection step by step, this guide demonstrates how to:

* Examine a basic MCP server with a simple `get_name` tool.
* Add FusionAuth OAuth configuration and token validation.
* Register an MCP client (such as Claude Desktop) and test the protected tool.

The following tutorial builds a working example of an OAuth-protected MCP server that you can extend with additional tools and security policies.

## Prerequisites

To follow this guide, you need the following installed:

* Docker Desktop (ensure it's running).
* An MCP client such as Claude Desktop or Cursor.

> * Python 3.12 or later (and `pip`).
> * Node.js version 20.18.1 or later (ensure it's available in your PATH).
> * FusionAuth 1.63.0 or later. The Docker stack in this guide uses the `latest` image tag, which satisfies this requirement. If you are running FusionAuth separately, version 1.63.0 or later is required because earlier versions do not advertise S256 PKCE support in the OIDC discovery document, which `mcp-remote` requires.

Reviewer: ```suggestion
* Python `3.12` or later (and `pip`).
* Node.js version `20.18.1` or later (ensure it's available in your PATH).
* FusionAuth `1.63.0` or later. Earlier versions do not advertise S256 PKCE support in the OIDC discovery document, which `mcp-remote` requires.
```
Suggestion: monospace versions to distinguish them from numbers; simplify the FusionAuth version explanation (hopefully it's obvious that `latest` satisfies this requirement... but maybe we should just set it to `1.67.0` since IIRC `latest` could use much older versions, if users haven't cleaned out their images lately?

<Aside type="tip">
Verify your Node.js installation:

```bash
node --version  # Should show v20.x.x or higher
npx --version   # Should show 10.x.x or higher
```
</Aside>

> You also need a **FusionAuth Essentials license** (or higher). Custom OAuth scopes require an Essentials license. If you already have a paid plan, you can find your license key at [account.fusionauth.io](https://account.fusionauth.io/account/plan/). Otherwise, [contact FusionAuth](/pricing) to obtain one.

Reviewer: IMO this needs to be added to the bulleted list above, not hidden beneath the node.js version check! We don't want users without this license to miss this and waste their time.

## Getting Started

Clone the repository and navigate to the starter code:

> ```bash
> git clone https://github.com/fusionauth/fusionauth-protected-mcp-server.git
> cd fusionauth-protected-mcp-server/unprotected-local-mcp
> ```

Reviewer: ````suggestion
```console
git clone https://github.com/fusionauth/fusionauth-protected-mcp-server.git
```

```console
cd fusionauth-protected-mcp-server/unprotected-local-mcp
```
````
suggestions: split this up for easier copy/paste, use the `console` language so it displays with a leading prompt to make it obvious that this is an interactive session, not a bash script. (this feedback applies to all interactive session snippets in this entire post)

<Aside type="note">

> This guide walks you through building OAuth protection step by step. If you prefer to skip ahead and see the completed code, navigate to the `protected-local-mcp` folder instead:

Reviewer: ```suggestion
This guide walks you through building OAuth protection. If you prefer to skip ahead and see the completed code, navigate to the `protected-local-mcp` folder instead:
```
suggestion: remove unnecessary phrase

> ```bash

Reviewer: ```suggestion
```console
```

cd fusionauth-protected-mcp-server/protected-local-mcp
```
</Aside>

### Configure Your FusionAuth License

Before starting the services, add your FusionAuth license key to the Kickstart configuration.

> Open `kickstart/kickstart.json` and replace the `licenseId` value on line 15:
> 
> ```json
> {
>   "variables": { ... },
>   "apiKeys": [ ... ],
>   "licenseId": "YOUR_LICENSE_KEY_HERE",
>   "requests": [ ... ]
> }
> ```
> 
> Replace `YOUR_LICENSE_KEY_HERE` with your actual FusionAuth license key.

Reviewer: ````suggestion
Open `kickstart/kickstart.json` and replace the `licenseId` value of `YOUR_LICENSE_KEY_HERE` with your actual FusionAuth license key:

```plaintext
{
  "variables": { ... },
  "apiKeys": [ ... ],
  "licenseId": "YOUR_LICENSE_KEY_HERE",
  "requests": [ ... ]
}
```

````
suggestions: switch to plaintext syntax highlighting, since the JSON highlighter _hates_ ellipses. Avoid explicitly pointing to "line 15" -- that will almost certainly change at some point, and I guarantee we'll forget to update this instruction. Instead of splitting up the instruction before and after the snippet, just provide the entire instruction before the snippet (especially odd when we say 'replace' before without saying what to replace the thing with!)

### Review The Project Structure

The project is structured as follows:

```
fusionauth-protected-mcp-server/
unprotected-local-mcp/          # Starter code (this guide begins here)
├── docker-compose.yml
├── kickstart/
│   └── kickstart.json
├── mcp-server/
│   ├── Dockerfile
│   ├── requirements.txt
│   └── server.py
└── setup/
    ├── requirements.txt
    └── setup_clients.py
protected-local-mcp/            # Completed code with OAuth protection
└── ...
protected-remote-mcp/           # Remote deployment variant
└── ...
```

## Understanding The Starter Code

Let's examine the starter MCP server before adding authentication.

### The Basic MCP Server

Open `mcp-server/server.py`:

<RemoteCode url={frontmatter.codeRoot + "/unprotected-local-mcp/mcp-server/server.py"} lang="python" tags="starter-server"/>

This is a minimal MCP server with:

* A single `get_name` tool that returns a hardcoded greeting.
* HTTP transport on port 8000.
* No authentication.

### Test The Unprotected Server

Start the Docker stack:

> ```bash

Reviewer: ```suggestion
```console
```

docker compose up -d
```

> Wait about 20 seconds for all services to start, including:

Reviewer: 20 seconds might be true on recent Macs if you have most or all of the images downloaded. It is very likely not true on an older machine, with mid-range internet speeds, especially if you lack some of the chonkier Docker images. I just tried this on my M4 Pro with 100Mbit download speeds, and it took several minutes. We should:
- get rid of an explicit time commitment
- explain how to tell when the services are all up and running, probably with example output

Something like:

```
This may take a few minutes. When the command finishes executing, you can proceed to the next step.
```

Step 5 of https://fusionauth.io/docs/get-started/quickstarts/web/express#install-and-run-fusionauth might be a good example to work from as well.

Starting with a simple, unprotected MCP server and adding OAuth protection step by step, this guide demonstrates how to:

* Examine a basic MCP server with a simple `get_name` tool
* Add FusionAuth OAuth configuration and token validation
* Register an MCP client (such as Claude Desktop) and test the protected tool

The following tutorial builds a working example of an OAuth-protected MCP server that you can extend with additional tools and security policies.

## Prerequisites

To follow this guide, you need the following installed:

* Docker Desktop (ensure it's running)
* Claude Desktop
* Python 3.12 or later (and `pip`)
* Node.js version 20.18.1 or later (ensure it's available in your PATH)

<Aside type="tip" nodark="true">
Verify your Node.js installation:

```bash
node --version  # Should show v20.x.x or higher
npx --version   # Should show 10.x.x or higher
```
</Aside>

> You also need a **FusionAuth Enterprise license**. Custom OAuth scopes require an Enterprise license. [Contact FusionAuth](/pricing) to obtain a license key.

Reviewer 2: If someone already has a paid plan, they can find their key in https://account.fusionauth.io/account/plan/

## Getting Started

Clone the starter repository:

```bash
git clone https://github.com/jamesdanielwhitford/fusionauth-protected-mcp-server.git
cd fusionauth-protected-mcp-server
```

<Aside type="note" nodark="true">
This guide walks you through building OAuth protection step by step. If you prefer to skip ahead and see the completed code, check out the `completed-app` branch:

```bash
git checkout completed-app
```
</Aside>

### Configure Your FusionAuth License

> Before starting the services, add your FusionAuth Enterprise license key to the Kickstart configuration.

Reviewer 3: Done, addressed in comment above.

[183814462](https://github.com/FusionAuth/fusionauth-site/commit/183814462)

  -H "Content-Type: application/json" \
  -H "Accept: application/json, text/event-stream" \
  -d '{
    "jsonrpc": "2.0",
    "id": 1,
    "method": "tools/call",
    "params": {
      "name": "get_name",
      "arguments": {}
    }
  }'
```

You should see:

```json
{
  "jsonrpc": "2.0",
  "id": 1,
  "result": {
    "content": [
      {
        "type": "text",
        "text": "Hello, World!"
      }
    ]
  }
}
```

The tool works, but anyone can call it because there's no authentication. Let's fix that.

### Verify That FusionAuth Is Running

Before you continue, confirm that FusionAuth started successfully. Check that you can access it:

```bash
curl http://localhost:9011/.well-known/openid-configuration | jq .issuer
```

You should see the following:

```json
"http://localhost:9011"
```

You can also visit http://localhost:9011 in your browser to log into the FusionAuth admin UI. Use the admin credentials created by the Kickstart configuration:

* **Email:** `admin@example.com`
* **Password:** `password`

Navigate to **Applications** to see the MCP Server application that the Kickstart configuration created.

## Adding OAuth Protection

> Add FusionAuth OAuth authentication piece by piece to protect the MCP server, starting with configuration and ending with a fully protected server.

Reviewer 2: ```suggestion
Now, let's add authentication piece by piece to protect the MCP server, starting with configuration and ending with a fully protected server.
```

### Step 1: Add Configuration And Imports

To configure the MCP server to connect to FusionAuth and import the OAuth components from FastMCP, replace the existing imports at the top of `server.py` with the following:

```python
import os
import logging
import json

import requests
from fastmcp import FastMCP
from fastmcp.server.auth import RemoteAuthProvider, AccessToken, TokenVerifier
from fastmcp.server.dependencies import get_access_token
from pydantic import AnyHttpUrl

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Add these configuration variables
FUSIONAUTH_URL = os.environ.get("FUSIONAUTH_URL", "http://fusionauth:9011")
FUSIONAUTH_EXTERNAL_URL = os.environ.get("FUSIONAUTH_EXTERNAL_URL", "http://localhost:9011")
MCP_SERVER_URL = os.environ.get("MCP_SERVER_URL", "http://localhost:8000")
MCP_APP_ID = os.environ.get("MCP_APP_ID", "e9fdb985-9173-4e01-9d73-ac2d60d1dc8e")
MCP_APP_SECRET = os.environ.get("MCP_APP_SECRET", "P6mR6aMjI-c7WJPe1RA95_S6HIpq9xzEqFGhaY5yqDc")
FUSIONAUTH_API_KEY = os.environ.get("FUSIONAUTH_API_KEY", "bf69486b-4733-4470-a592-f1bfce7af580")
```

In this code:

> * `FUSIONAUTH_URL` adds the internal Docker network address (`http://fusionauth:9011`), which the MCP server container uses to reach FusionAuth.

Reviewer 2: you only have 4 variables here, but you added 6 above. Why is that?

### Step 2: Implement Token Validation

After the configuration variables, extend the FastMCP `TokenVerifier` base class to create the following token verifier:

```python
class FusionAuthTokenVerifier(TokenVerifier):
    """Verifies tokens by calling FusionAuth's UserInfo endpoint."""

    def __init__(
        self,
        fusionauth_url: str,
        client_id: str,
        client_secret: str | None = None,
        required_scopes: list[str] | None = None,
    ):
        super().__init__(required_scopes=required_scopes)
        self.fusionauth_url = fusionauth_url
        self.client_id = client_id
        self.client_secret = client_secret

    async def verify_token(self, token: str) -> AccessToken | None:
        try:
            # Validate token via UserInfo endpoint
            resp = requests.get(
                f"{self.fusionauth_url}/oauth2/userinfo",
                headers={"Authorization": f"Bearer {token}"},
            )

            if resp.status_code != 200:
                logger.warning("Token validation failed")
                return None

            userinfo = resp.json()

            # Decode token claims

>             import base64

Reviewer 3: The UserInfo endpoint doesn't return scopes, and neither does the introspection endpoint. 

FastMCP checks `AccessToken.scopes` internally to gate tool access, so we have to decode the JWT to extract them. We verified this: passing `scopes=[]` results in `403 insufficient_scope` even with a valid token. Added a comment in the code explaining this.

```diff
- # Decode token claims
+ # The UserInfo endpoint does not return scopes, so we decode the JWT
+ # to extract them. FastMCP uses scopes to enforce access control on
+ # individual tools.
```

[19ad541](https://github.com/FusionAuth/fusionauth-site/commit/19ad541ea)

```python
# Create token verifier with required scopes
token_verifier = FusionAuthTokenVerifier(
    fusionauth_url=FUSIONAUTH_URL,
    client_id=MCP_APP_ID,
    client_secret=MCP_APP_SECRET,
    required_scopes=["get_name"],
)

# Configure OAuth provider
auth = RemoteAuthProvider(
    token_verifier=token_verifier,
    authorization_servers=[AnyHttpUrl(FUSIONAUTH_EXTERNAL_URL)],
    base_url=MCP_SERVER_URL,
)

# Create MCP server with OAuth protection
mcp = FastMCP(
    name="FusionAuth MCP Server",
    auth=auth,
)
```

This updated initialization does the following:

* **Creates a token verifier instance with required scopes:** `required_scopes=["get_name"]` ensures that only tokens with the `get_name` scope can execute tools.
* **Wraps the instance in `RemoteAuthProvider` (FastMCP's OAuth integration):** `RemoteAuthProvider` integrates OAuth with FastMCP's HTTP transport.
* **Enables MCP clients to discover OAuth endpoints:** `authorization_servers` points to FusionAuth's external URL, where MCP clients can discover the endpoints.
* **Passes the auth provider to FastMCP:** `auth=auth` enables authentication on the MCP server.

> Now all tool calls require a valid access token with the `get_name` scope.

Reviewer 2: What happens if the access token is not provided? I think a 401?

### Step 4: Access Authenticated User Information

With authentication enabled, you can now access information about the authenticated user. FastMCP provides a `get_access_token()` dependency that gives you the validated token in your tools.

Replace the `get_name` tool with this implementation:

```python

> def _lookup_user_name(user_id: str) -> str:

Reviewer 2: As mentioned above, this can be done with the Userinfo endpoint response. Any reason not to do that?

(Then you don't need the FusionAuth API key.)

```bash
docker compose up -d --build mcp-server
```

Wait a few seconds for the container to restart. The MCP server now requires OAuth authentication.

## Testing OAuth Protection

Let's verify that the authentication works.

### Test Without A Token

Try calling the tool without authentication:

```bash
curl -X POST http://localhost:8000/mcp \
  -H "Content-Type: application/json" \
  -H "Accept: application/json, text/event-stream" \
  -d '{
    "jsonrpc": "2.0",
    "id": 1,
    "method": "tools/call",
    "params": {
      "name": "get_name",
      "arguments": {}
    }
  }'
```

You should see an authentication error, indicating that the tool is now protected.

### Check That Authentication Is Enabled

Verify that the server is running and authentication is active:

```bash
docker compose logs mcp-server | grep "Uvicorn running"
```

You should see output confirming the server started successfully on port 8000.

## Registering MCP Clients As OAuth Clients

Before MCP clients like Claude Desktop or Cursor can connect to your protected MCP server, you need to register them as OAuth clients in FusionAuth.

### The Registration Process

> MCP clients need their own OAuth application in FusionAuth, separate from the MCP server application. When an MCP client initiates the OAuth flow, it looks like this:

Reviewer 2: Each MCP client needs its own application, correct? If so, that is worth noting.

1. The MCP client redirects the user to FusionAuth to log in.
2. The user authenticates their account and grants consent.
3. FusionAuth redirects the user back to the MCP client with an authorization code.
4. The MCP client exchanges the code for an access token.
5. The MCP client passes the token to the MCP server with each request.

For the OAuth flow to work, FusionAuth needs to know:

* Which redirect URLs are allowed for each MCP client
* Whether PKCE is required (some MCP clients use PKCE for security)
* Which OAuth grants are enabled

### What The Setup Script Does

The `setup_clients.py` script automates this registration process. Let's look at the key parts.

The script defines the client configuration with redirect URLs:

```python
CLIENT_CONFIGS = {
    "claude_desktop": {
        "name": "Claude Desktop",
        "redirect_urls": [

>             "http://localhost:3334/oauth/callback",

Reviewer 2: How did you find out the `3334` port number? Is this documented someplace? Trial and error? Is it the same for all claude desktop clients?

Same for 16442?

Reviewer 3: Switched to wildcard redirect URLs (`http://localhost:*/oauth/callback`) with `authorizedURLValidationPolicy: AllowWildcards` in FusionAuth, so the OAuth callback works on any port regardless of which client is used. This avoids hardcoding port numbers that vary by client and run.

```diff
- "http://localhost:3334/oauth/callback",
- "http://127.0.0.1:3334/oauth/callback",
- "http://localhost:16442/oauth/callback",
- "http://127.0.0.1:16442/oauth/callback",
+ "http://localhost:*/oauth/callback",
+ "http://127.0.0.1:*/oauth/callback",
```

[85af3ce](https://github.com/FusionAuth/fusionauth-site/commit/85af3cead)

        "application": {
            "name": config["name"],
            "oauthConfiguration": {
                "authorizedRedirectURLs": config["redirect_urls"],
                "clientAuthenticationPolicy": "NotRequiredWhenUsingPKCE",
                "enabledGrants": ["authorization_code", "refresh_token"],
                "generateRefreshTokens": True,
                "proofKeyForCodeExchangePolicy": "Required",
                "requireClientAuthentication": False,
                "scopeHandlingPolicy": "Compatibility",
                "unknownScopePolicy": "Allow",
            },
        }
    }

    resp = requests.post(
        f"{base_url}/api/application/{app_id}",
        headers={"Authorization": api_key, "Content-Type": "application/json"},
        json=body,
    )
```

This includes the following key configuration values:

* `proofKeyForCodeExchangePolicy: "Required"` enforces PKCE for security
* `clientAuthenticationPolicy: "NotRequiredWhenUsingPKCE"` means that no client secret is needed when PKCE is used
* `scopeHandlingPolicy: "Compatibility"` allows custom scopes like `get_name`
* `unknownScopePolicy: "Allow"` permits scopes that aren't pre-defined in FusionAuth

> The script returns the generated client ID, which you need for configuring Claude Desktop.

Reviewer 2: ```suggestion
The script returns the generated client Id, which you need for configuring Claude Desktop.
```

### Run The Client Registration Script

Install the dependencies and run the setup script:

```bash
cd setup
pip install -r requirements.txt
python setup_clients.py
```

The script prompts you to select which clients to register. Choose the client(s) you want to use:

```
Available MCP clients:
  1. Claude Desktop
  2. Cursor
  3. All

Select clients to register (comma-separated numbers): 1
```

You should see output like this:

```
Registering 1 client(s)...

  Creating Claude Desktop...
  Created Claude Desktop (Client ID: 9436b02d-9305-4772-a237-5fabcf6e7f6a)

========================================
Setup complete!
========================================

  MCP configuration for Claude Desktop:
  Client ID: 9436b02d-9305-4772-a237-5fabcf6e7f6a
```

> Copy the client ID you need for the next step. You can also verify the registration in the FusionAuth admin UI by navigating to **Applications** to see the new client alongside the MCP Server application.

Reviewer 2: Also, use a `<Breadcrumb>` here for the Applications tab.

<Aside type="note" nodark="true">
You can register multiple clients by selecting option 3 (All) or by providing comma-separated numbers (for example, `1,2`). Each client gets its own application and client ID in FusionAuth.
</Aside>

## Configuring Your MCP Client

Now let's configure the MCP client to connect to the OAuth-protected MCP server. The configuration is the same for all clients, but the setup process differs slightly.

### Configuration Format

All MCP clients use the same configuration structure:

```json
{
  "mcpServers": {
    "fusionauth-mcp": {
      "command": "npx",
      "args": [
        "mcp-remote",
        "http://localhost:8000/mcp",
        "--allow-http",
        "--static-oauth-client-info",
        "{\"client_id\":\"YOUR_CLIENT_ID_HERE\"}"
      ]
    }
  }
}
```

Replace `YOUR_CLIENT_ID_HERE` with the client ID from the registration script (the one that corresponds to your MCP client).

Each parameter does the following:

* `command` uses `npx` to run the `mcp-remote` bridge tool.

> * `mcp-remote` bridges MCP clients (which expect stdio) to your HTTP-based MCP server.

Reviewer 2: Like if I were to run the MCP server on an EC2 instance with TLS support, we wouldn't use STDIO, would we?

* `--allow-http` allows non-HTTPS connections for local development.
* `--static-oauth-client-info` provides the preregistered client ID from the setup script.

The `mcp-remote` tool is part of the official MCP ecosystem and handles the OAuth flow automatically.

### For Claude Desktop

1. Open Claude Desktop.

> 2. Navigate to **Settings → Developer**.

Reviewer 2: Use breadcrumb here.

6. **Validation:** The MCP server validates tokens via FusionAuth's UserInfo endpoint before executing tools.

### Token Validation Approach

The MCP server validates tokens using the FusionAuth UserInfo endpoint. This is a standard OpenID Connect endpoint that:

* Accepts Bearer tokens in the Authorization header
* Returns user information if the token is valid
* Does not require client authentication (because the token authenticates the request)

If the UserInfo endpoint returns `200 OK`, the token is valid and active.

### Scope Enforcement

The `required_scopes=["get_name"]` parameter ensures that:

* Only tokens with the `get_name` scope can execute the tool
* FastMCP automatically validates scopes before tool execution
* Users must explicitly grant this scope during the consent flow

### Custom Scopes In FusionAuth

FusionAuth requires an Enterprise license to use custom scopes. The Kickstart configuration:

1. Activates the license
2. Creates the `get_name` scope
3. Configures the application with the scope

When users authorize the application, FusionAuth redirects them back to the MCP client with an "Authorization successful!" message.

## Production Considerations

> This example demonstrates the core patterns for OAuth-protected MCP servers. For production deployments, consider including:

Reviewer 2: Also, cleaning up clients who haven't logged in for a while. Turn on self service registration in kickstart so we can reap applications that haven't been logged into in a while by looking at the login records ( https://fusionauth.io/docs/apis/login#search-login-records ). 

Also, maybe a nod to how to preregister clients at scale.

```bash
curl http://localhost:9011/.well-known/openid-configuration
```

### The Token Validation Fails

Check the MCP server logs:

```bash
docker compose logs mcp-server
```

Then verify that the user is registered to the MCP client application and that the token includes the required scopes.

### Your MCP Client Doesn't Show The Server

Verify that your configuration JSON syntax is valid, restart the client after configuration changes, and confirm the client ID matches what the setup script created.

### The OAuth Flow Completes, But Tools Aren't Available

Completely quit and restart your MCP client (such as Claude Desktop or Cursor) after the OAuth flow completes. The client needs to restart to pick up the newly obtained access token.

### The "Invalid redirect_uri" Error Occurs During OAuth

OAuth clients redirect users back to a specific URL after authentication. If FusionAuth doesn't recognize the redirect URL, it rejects the request. When you see this error, look at the error message to find which URL was attempted (for example, `http://localhost:16442/oauth/callback`). Then add that URL to your client application's authorized redirect URLs in FusionAuth.

> You can fix this in two ways:

Reviewer 3: Done. Removed the API approach and kept only the Admin UI method.

```diff
- You can fix this in two ways:
- 1. **Via the FusionAuth Admin UI:** Navigate to **Applications → [Your Client] → OAuth tab → Authorized redirect URLs**, then add the missing URL.
- 2. **Via the API:** Update the application with a `PATCH` request: ...
+ To fix this, navigate to <Breadcrumb>Applications -> [Your Client] -> OAuth -> Authorized redirect URLs</Breadcrumb> in the FusionAuth admin UI and add the missing URL.
```

[183814462](https://github.com/FusionAuth/fusionauth-site/commit/183814462)

1. The MCP client redirects the user to FusionAuth to log in.
2. The user authenticates their account and grants consent.
3. FusionAuth redirects the user back to the MCP client with an authorization code.
4. The MCP client exchanges the code for an access token.
5. The MCP client passes the token to the MCP server with each request.

For the OAuth grant to work, FusionAuth needs to know:

* Which redirect URLs are allowed for each MCP client
* Whether PKCE is required (some MCP clients use PKCE for security)
* Which OAuth grants are enabled

### What The Setup Script Does

The `setup_clients.py` script automates this registration process. Let's look at the key parts.

The script prompts you for a name for the client, then creates an OAuth application in FusionAuth. Run it once for each MCP client you want to register. It uses wildcard redirect URLs so the OAuth callback works on any port:

```python

> REDIRECT_URLS = [

Reviewer 2: Hmmm. The wildcard is a pretty big security risk, because it means anyone can stand up a process listening on a port and conceivably get the user to click on a link and get an authorization code.

I think I asked about how the port numbers were determined--are they documented someplace or did you discover them via experimentation.

from fastmcp.server.auth import RemoteAuthProvider, AccessToken, TokenVerifier
from fastmcp.server.dependencies import get_access_token
from pydantic import AnyHttpUrl

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Add these configuration variables
FUSIONAUTH_URL = os.environ.get("FUSIONAUTH_URL", "http://fusionauth:9011")
FUSIONAUTH_EXTERNAL_URL = os.environ.get("FUSIONAUTH_EXTERNAL_URL", "http://localhost:9011")
MCP_SERVER_URL = os.environ.get("MCP_SERVER_URL", "http://localhost:8000")
MCP_APP_ID = os.environ.get("MCP_APP_ID", "e9fdb985-9173-4e01-9d73-ac2d60d1dc8e")
MCP_APP_SECRET = os.environ.get("MCP_APP_SECRET", "P6mR6aMjI-c7WJPe1RA95_S6HIpq9xzEqFGhaY5yqDc")
```

In this code:

* `FUSIONAUTH_URL` adds the internal Docker network address (`http://fusionauth:9011`), which the MCP server container uses to reach FusionAuth.
* `FUSIONAUTH_EXTERNAL_URL` adds the public localhost URL (`http://localhost:9011`), which browsers use during the OAuth flow.
* `MCP_SERVER_URL` adds the public URL of the MCP server, used by FastMCP to construct OAuth redirect URLs.
* `MCP_APP_ID` adds the application ID for the MCP server application.

> * `MCP_APP_SECRET` is a placeholder client secret. Token validation uses the UserInfo endpoint, which requires no client authentication, so this value is not used at runtime.

Reviewer 2: I see this variable used elsewhere. If it is not required, should we remove it?

The values of the configuration variables match the Kickstart configuration in `kickstart/kickstart.json`. When you ran `docker compose up`, FusionAuth detected a fresh installation and automatically ran the Kickstart configuration, which created:

* An OAuth application called "MCP Server"
* A test user (`test@example.com`/`password`)
* An admin user (`admin@example.com`/`password`) with access to the FusionAuth admin UI
* An API key
* A custom `get_name` scope

The Ids and keys are hardcoded in the Kickstart file. In production, you'd use environment variables or secrets management instead of hardcoded values.

### Step 2: Implement Token Validation

> After the configuration variables, extend the FastMCP `TokenVerifier` base class to create the following token verifier:

Reviewer 2: ```suggestion
After the configuration variables, add this code to extend the FastMCP `TokenVerifier` base class. You are creating a token verifier, which ensures that the MCP server responds only to properly authorized requests:
```

> ```python

Reviewer 3: FastMCP's built-in JWT verifier would have worked. 

We've since replaced that with `validate_jwt` from the FusionAuth Python client (as requested in the other comments), which does the same thing. 

It verifies the token signature against FusionAuth's public keys and returns the claims including `scope`, but through the FusionAuth library rather than FastMCP's JWT.

class FusionAuthTokenVerifier(TokenVerifier):
    """Verifies tokens by calling FusionAuth's UserInfo endpoint."""

    def __init__(
        self,
        fusionauth_url: str,
        client_id: str,
        client_secret: str | None = None,
        required_scopes: list[str] | None = None,
    ):
        super().__init__(required_scopes=required_scopes)
        self.fusionauth_url = fusionauth_url
        self.client_id = client_id
        self.client_secret = client_secret

    async def verify_token(self, token: str) -> AccessToken | None:
        try:
            # Validate token via UserInfo endpoint. With the profile, email, and
            # openid scopes requested, the response includes user profile data
            # (name, email) so no separate API call is needed.

>             resp = requests.get(

Reviewer 3: Done. The UserInfo call now uses `client.retrieve_user_info_from_access_token(access_token.token)` from `FusionAuthClient`. The `requests` library has been removed entirely.

Commit: [a8458fc](https://github.com/ritza-co/fusionauth-protected-mcp-server/commit/a8458fc77a402a9f8654da4c75c3e3e51b067a89)

                return None

            userinfo = resp.json()

            # The UserInfo endpoint does not return scopes, so we decode the JWT
            # to extract them. FastMCP uses scopes to enforce access control on
            # individual tools.
            import base64
            payload_b64 = token.split('.')[1]
            payload_b64 += '=' * (4 - len(payload_b64) % 4)  # Add padding
            claims = json.loads(base64.urlsafe_b64decode(payload_b64))

            scopes = claims.get("scope", "").split() if claims.get("scope") else []

            # Merge userinfo into claims so tools can access profile data
            # (e.g. name, email) without needing a separate API call.
            claims.update(userinfo)

            return AccessToken(
                token=token,
                client_id=claims.get("sub", ""),
                scopes=scopes,
                expires_at=claims.get("exp"),
                claims=claims,
            )
        except Exception as e:
            logger.error("Failed to validate token: %s", e)
            return None
```

This token verifier validates access tokens by:

> 1. Calling FusionAuth's UserInfo endpoint with the token. A `200 OK` response confirms the token is valid

Reviewer 3: Bullet point edits made here: [00e0ed4](https://github.com/FusionAuth/fusionauth-site/pull/4115/commits/00e0ed4fe9f72879a13abd7185e17245794fdb7f)

)

# Configure OAuth provider
auth = RemoteAuthProvider(
    token_verifier=token_verifier,
    authorization_servers=[AnyHttpUrl(FUSIONAUTH_EXTERNAL_URL)],
    base_url=MCP_SERVER_URL,
)

# Create MCP server with OAuth protection
mcp = FastMCP(
    name="FusionAuth MCP Server",
    auth=auth,
)
```

This updated initialization does the following:

* **Creates a token verifier instance with required scopes:** `required_scopes=["openid", "profile", "email", "get_name"]` ensures that only tokens with all four scopes can execute tools. The `openid`, `profile`, and `email` scopes tell FusionAuth to include user profile data in the UserInfo response. The `get_name` scope is the custom scope that gates access to the `get_name` tool.
* **Wraps the instance in `RemoteAuthProvider` (FastMCP's OAuth integration):** `RemoteAuthProvider` integrates OAuth with FastMCP's HTTP transport.
* **Enables MCP clients to discover OAuth endpoints:** `authorization_servers` points to FusionAuth's external URL, where MCP clients can discover the endpoints.
* **Passes the auth provider to FastMCP:** `auth=auth` enables authentication on the MCP server.

> Now all tool calls made to this MCP server require a valid access token with the `get_name` scope. If no token is provided, the server returns a `401 Unauthorized` response.

Reviewer 2: just the `get_name` scope?

```

You should see output confirming the server started successfully on port 8000.

## Registering MCP Clients As OAuth Clients

Before MCP clients like Claude Desktop or Cursor can connect to your protected MCP server, you need to register them as OAuth clients in FusionAuth.

### The Registration Process

Each MCP client needs its own dedicated OAuth application in FusionAuth, separate from the MCP server application. This means if you want to support both Claude Desktop and Cursor, you register two separate applications, one per client. When an MCP client initiates the OAuth grant, it looks like this:

1. The MCP client redirects the user to FusionAuth to log in.
2. The user authenticates their account and grants consent.
3. FusionAuth redirects the user back to the MCP client with an authorization code.
4. The MCP client exchanges the code for an access token.
5. The MCP client passes the token to the MCP server with each request.

For the OAuth grant to work, FusionAuth needs to know:

* Which redirect URLs are allowed for each MCP client
* Whether PKCE is required (some MCP clients use PKCE for security)
* Which OAuth grants are enabled

### What The Setup Script Does

> The `setup_clients.py` script automates this registration process. Let's look at the key parts.

Reviewer 2: ```suggestion
The `setup_clients.py` script automates this registration process. It is located in the `setup` directory. Let's look at the key parts.
```

* `mcp-remote` bridges MCP clients (which expect stdio) to your HTTP-based MCP server. MCP client config files only support stdio servers, so `mcp-remote` is required to connect to any HTTP-based MCP server, regardless of whether auth is involved.
* `--allow-http` allows non-HTTPS connections for local development.
* `--static-oauth-client-info` provides the preregistered client Id from the setup script. This is required because FusionAuth does not support Dynamic Client Registration (DCR). Native MCP clients that attempt DCR will fail to connect. By passing a pre-registered client Id, `mcp-remote` bypasses DCR entirely.

### For Claude Desktop

1. Open Claude Desktop.
2. Navigate to <Breadcrumb>Settings -> Developer</Breadcrumb>.
3. Click <InlineUIElement>Edit Config</InlineUIElement>.
4. Add the configuration above.
5. Save the file.

### For Cursor

1. Press `Cmd + Shift + P` (macOS) or `Ctrl + Shift + P` (Windows and Linux).
2. Type `MCP` and select <InlineUIElement>Open MCP Settings</InlineUIElement>.
3. Click <InlineUIElement>New MCP Server</InlineUIElement>.
4. Add the configuration above.
5. Save the file.

## Testing The Protected Server

> Restart your MCP client to load the new configuration.

Reviewer 2: By the way, this is going to be fixed soon, it is issue https://github.com/fusionAuth/fusionauth-issues/issues/1177

> Restart your MCP client to load the new configuration.

Reviewer 2: ```
$ npm view mcp-remote version

0.1.38
```

1. **Discovery:** The MCP client discovers OAuth endpoints from FusionAuth's OpenID Connect configuration.
2. **Authorization:** The MCP client redirects the user's browser to FusionAuth's login page.
3. **Authentication:** The user logs in and grants consent for the requested scopes.
4. **Token exchange:** FusionAuth returns an authorization code, which the MCP client exchanges for an access token.
5. **API calls:** The MCP client includes the access token in requests to the MCP server.
6. **Validation:** The MCP server validates tokens via FusionAuth's UserInfo endpoint before executing tools.

### Token Validation Approach

The MCP server validates tokens using the FusionAuth UserInfo endpoint. This is a standard OpenID Connect endpoint that:

* Accepts Bearer tokens in the Authorization header
* Returns user information if the token is valid
* Does not require client authentication (because the token authenticates the request)

If the UserInfo endpoint returns `200 OK`, the token is valid and active.

### Scope Enforcement

The `required_scopes=["openid", "profile", "email", "get_name"]` parameter ensures that:

> * Only tokens containing all four scopes can execute tools

Reviewer 3: Bullet point edits made here: [00e0ed4](https://github.com/FusionAuth/fusionauth-site/pull/4115/commits/00e0ed4fe9f72879a13abd7185e17245794fdb7f)

* FastMCP automatically validates scopes before tool execution
* Users must explicitly grant these scopes during the consent flow
* The `openid`, `profile`, and `email` scopes cause FusionAuth to include user profile data in the UserInfo response, so the MCP server can read the user's name without a separate API call

### Custom Scopes In FusionAuth

FusionAuth requires an Essentials license (or higher) to use custom scopes. The Kickstart configuration:

> 1. Activates the license

Reviewer 2: Same comment about bullet point lists. Either have them be full sentences or don't start them with capitals.

2. Creates the `get_name` scope
3. Configures the application with the scope

When users authorize the application, FusionAuth redirects them back to the MCP client with an "Authorization successful!" message.

## Production Considerations

This example demonstrates the core patterns for OAuth-protected MCP servers. For production deployments, consider including:

* **HTTPS:** Use HTTPS for all endpoints, remove the `--allow-http` flag, and configure proper TLS certificates.
* **Token caching:** Cache UserInfo responses to avoid validating tokens on every request. Set the cache TTL to match token expiration.

> * **Refresh tokens:** Configure your FusionAuth application to issue refresh tokens for long-lived sessions.

Reviewer 2: Do MCP clients handle the refresh grant?
